### PR TITLE
WIP: Automatic reloading of backend on config changes.

### PR DESCRIPTION
**Pull request by eskimor**
Not yet tested at all. But as this is already lying around for a while and I almost lost it, thought I'll put it up already.

Problem this PR is trying to solve:

If you do `ob deploy push` with only config file changes (no new version of the app gets deployed), the backend will not pick up the new configs. (You would need to do a manual `systemctl restart backend`.) With this PR, the backend should be restarted automatically on config file changes.

### CODE REVIEW

**Comment by ryantrinkle** [comment]
I like the comprehensiveness here, but can't this result in the backend being restarted twice during a single deploy?

File: default.nix [hunk 1]
```diff
@@ -191,46 +191,64 @@ in rec {
...
-          description = "${user} service";
-          home = "/var/lib/${user}";
-          createHome = true;
-          isSystemUser = true;
-          group = group;
+        # Restart backend on configuration changes:
+        systemd.services.restart-${name} = {
+          serviceConfig = {
+            type = "oneshot";
+            ExecStart="${nixpkgs.systemd.out}/bin/systemctl restart ${name}.service";
```
Comment (3noch) [inline default.nix]: Should probably be `${config.systemd.package}/bin/systemctl`

**Comment by 3noch** [comment]
Have you explored a more "stateless" approach where you just piggyback off of `restartIfChanged` using some sort of hash of the config contents?

**Comment by eskimor** [comment]
> I like the comprehensiveness of this approach, but can't this result in the backend being restarted twice during a single deploy?

I think, you are right. I actually tried to avoid exactly that, but somehow I did not think this through apparently. Not only does it restart the backend twice, it would do so in quite a bad way:

- Once when the configs are updated
- Right afterwards at profile switch

Problem: At the first reload, the old backend would be restarted, but with already new config files, which might not be parse-able by the old backend.

So auto-reloading is probably a bad idea in general. It would be better to just reload on deployment in case there was no profile switch.

**Comment by 3noch** [comment]
@roberth Couldn't `ob deploy` calculate a hash of the config and pass it as an argument to the build? Then you could do something as silly as adding `# config hash = ${configHash}` to the systemd service and it will suddenly get a different path and restart.

**Comment by eskimor** [comment]
@3noch That's a really smart idea, solving both problems at once. I like that! :-)

**Comment by 3noch** [comment]
c.f. #372

**Comment by ryantrinkle** [comment]
@3noch I think this is approximately a good idea, but what happens if the hash gets out of sync with the config?  We can't just put the config into the nix store, but we also should design this so that they can't get out of sync.

**Comment by 3noch** [comment]
By "get out of sync" do you mean "someone ssh'ed into the server and tweaked configs manually"?

**Comment by ryantrinkle** [comment]
@3noch No, although that is a use case we need to support.

Here's what we should be shooting for:

- if `ob deploy push` succeeds, at the end of that process, the backend is running with the correct config
- if `ob deploy push` fails, it doesn't leave things in a half-deployed state

**Comment by 3noch** [comment]
In that case I'm not sure what else would cause config to get out of sync? If, barring manual mutation on the server, the only way for configs to be changed is with `ob deploy` then the hash will be calculated every time and the configs copied every time.

**Comment by 3noch** [comment]
Maybe I should clarify that slapping a comment in the derivation wasn't actually the *end* of the idea. What would need to be done, in reality, is have separate directories for each hash-fingerprinted version of the config. Then the backend would symlink to the appropriate one. That is just replicating nix behavior but not in /nix/store

**Comment by ryantrinkle** [comment]
After some more thought, I think if an `ob deploy push` fails partway through, we're OK anyway: the way to recover will be to do another `ob deploy push`, at which point the config will be fixed.

**Comment by madeline-os** [comment]
Closing in favor of #940 (which was merged. yay)

**Pull request closed without merging**